### PR TITLE
Ship InMemoryConfigProvider 

### LINKED ISSUE #1713: Ship InMemoryConfigProvider
https://github.com/microsoft/reverse-proxy/blob/main/testassets/ReverseProxy.Code/InMemoryConfigProvider.cs

InMemoryConfigProvider is convenient for providing config from custom sources. It's copied several times in tests and samples, and even in customer projects.
https://github.com/DuendeSoftware/BFF/blob/main/src/Duende.Bff.Yarp/InMemoryConfigProvider.cs

The main reason I haven't done this yet is because I haven't implemented an update pattern.

### PR DESCRIPTION
**Pull request by Tratcher**
Fixes #1713

This ships InMemoryConfigProvider, removes redundant copies, and proposes an update pattern.

### CODE REVIEW

File: samples/ReverseProxy.Code.Sample/Startup.cs [hunk 1]
```diff
@@ -58,7 +62,7 @@ private RouteConfig[] GetRoutes()
             {
                 new RouteConfig()
                 {
-                    RouteId = "route1",
+                    RouteId = "route" + new Random().Next(), // Forces a new route id each time GetRoutes is called.
```
Comment (MihaZupan) [inline samples/ReverseProxy.Code.Sample/Startup.cs:65]:
```suggestion
                    RouteId = "route" + Random.Shared.Next(), // Forces a new route id each time GetRoutes is called.
```

File: src/ReverseProxy/Configuration/InMemoryConfigProviderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+public static class InMemoryConfigProviderExtensions
+{
+    /// <summary>
+    /// Adds an InMemoryConfigProvider
+    /// </summary>
+    public static IReverseProxyBuilder LoadFromMemory(this IReverseProxyBuilder builder, IReadOnlyList<RouteConfig> routes, IReadOnlyList<ClusterConfig> clusters)
+    {
+        builder.Services.AddSingleton(new InMemoryConfigProvider(routes, clusters));
+        builder.Services.AddSingleton<IProxyConfigProvider>(s => s.GetRequiredService<InMemoryConfigProvider>());
```
Comment (MihaZupan) [inline src/ReverseProxy/Configuration/InMemoryConfigProviderExtensions.cs:17]:
```suggestion
        builder.Services.AddSingleton<IProxyConfigProvider>(new InMemoryConfigProvider(routes, clusters));
```

Isn't this equivalent?

File: src/ReverseProxy/Configuration/InMemoryConfigProvider.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.Extensions.Primitives;
+
+namespace Yarp.ReverseProxy.Configuration;
+
+/// <summary>
+/// Provides an implementation of IProxyConfigProvider to support config being generated by code.
+/// </summary>
+public class InMemoryConfigProvider : IProxyConfigProvider
```
Comment (MihaZupan) [inline src/ReverseProxy/Configuration/InMemoryConfigProvider.cs:13]: Should we start out with it being sealed? `Update` isn't virtual so deriving from it has limited value as-is.

**Review by MihaZupan — APPROVED** [review]

File: src/ReverseProxy/Configuration/InMemoryConfigProviderExtensions.cs (hunk 2 above)
Comment (Tratcher) [reply src/ReverseProxy/Configuration/InMemoryConfigProviderExtensions.cs:17]: InMemoryConfigProvider is intentionally registered independently so you can resolve InMemoryConfigProvider without downcasting.

**Review by greenEkatherine — APPROVED** [review]

**Pull request merged**